lib/SILOptimizer/Mandatory/SILGenCleanup.cpp

```diff
@@ -104,39 +108,219 @@ struct SILGenCleanup : SILModuleTransform {
   void run() override;
 
   bool completeOSSALifetimes(SILFunction *function);
+  template <typename Range>
+  bool completeLifetimesInRange(Range const &range,
+                                OSSALifetimeCompletion &completion,
+                                BasicBlockSet &completed);
 };
 
+// Iterate over `iterator` until finding a block in `include` and not in
+// `exclude`.
+SILBasicBlock *findFirstBlock(SILFunction::iterator &iterator,
+                              BasicBlockSet const &include,
+                              BasicBlockSet const &exclude) {
+  while (iterator != include.getFunction()->end()) {
+    auto *block = &*iterator;
+    iterator = std::next(iterator);
+    if (!include.contains(block))
+      continue;
+    if (exclude.contains(block))
+      continue;
+    return block;
+  }
+  return nullptr;
+}
+
+// Walk backward from `from` following first predecessors until finding the
+// first already-reached block.
+SILBasicBlock *findFirstLoop(SILFunction *function, SILBasicBlock *from) {
+  BasicBlockSet path(function);
+  auto *current = from;
+  while (auto *block = current) {
+    current = nullptr;
+    if (!path.insert(block)) {
+      return block;
+    }
+    if (block->pred_empty()) {
+      return nullptr;
+    }
+    current = *block->getPredecessorBlocks().begin();
+  }
+  llvm_unreachable("finished function-exiting loop!?");
+}
+
+/// Populate `roots` with the last blocks that are discovered via backwards
+/// walks along any non-repeating paths starting at the ends in `walks`.
+void collectReachableRoots(SILFunction *function, BasicBlockWorklist &backward,
+                           StackList<SILBasicBlock *> &roots) {
+  assert(!backward.empty());
+  assert(roots.empty());
+
+  // Always include the entry block as a root.  Currently SILGen will emit
+  // consumes in unreachable blocks of values defined in reachable blocks (e.g.
+  // test/SILGen/unreachable_code.swift:testUnreachableCatchClause).
+  // TODO: [fix_silgen_destroy_unreachable] Fix SILGen not to emit such
+  //                                        destroys and don't add the entry
+  //                                        block to roots here.
+  roots.push_back(function->getEntryBlock());
+
+  // First, find all blocks backwards-reachable from dead-end blocks.
+  while (auto *block = backward.pop()) {
+    for (auto *predecessor : block->getPredecessorBlocks()) {
+      backward.pushIfNotVisited(predecessor);
+    }
+  }
+
+  // Simple case: unpredecessored blocks.
+  //
+  // Every unpredecessored block reachable from some dead-end block is a root.
+  for (auto &block : *function) {
+    if (&block == function->getEntryBlock()) {
```

Comment:
You don't need this condition because the entry block has no predecessors (which is checked below)

Reply:
Not quite.  The code here is
```
    if (&block == function->getEntryBlock()) {
      continue;
    }
    if (!block.pred_empty())
      continue;
```
The entry block has no predecessors, so we don't continue below.